lib/Sema/TypeCheckMacros.cpp

```diff
@@ -498,3 +498,165 @@ Expr *swift::expandMacroExpr(
          "Type checking changed the result type?");
   return expandedExpr;
 }
+
+/// Expands the given macro expansion declaration and type-check the result.
+bool swift::expandFreestandingDeclarationMacro(
```

Comment:
Hmm, lots of duplication with the `expandMacro` for expressions. Can we unify them somehow?

Reply:
We can definitely pull out the buffer creation, source info creation and the call to ASTGen in a follow-up patch.